File: testcases/preorder/preorder.lua
import("array")

function preorder(node)
  function inner(node, arr) 
    if (node == null) then 
      return arr 
	  end
    array.append(arr, node.value)
    inner(node.left, arr)
    inner(node.right, arr)
    return arr
  end

  return inner(node, {})
end

return preorder

File: testcases/preorder/preorder.js
const testCase = require('../test_fn.js');
const deserialize = require('../deserialize.js');

function preorder(node, result = []) {
  if (!node) return result;
  result.push(node.value);
  preorder(node.left, result);
  preorder(node.right, result);
  return result;
}

testCase([deserialize([1, 2, 3, 4, 5, 6, 7])], preorder);
